mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/dialect/impl/EmbeddableStateFinder.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Hibernate OGM, Domain model persistence for NoSQL datastores
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+
+package org.hibernate.ogm.datastore.mongodb.dialect.impl;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.hibernate.ogm.dialect.spi.TupleContext;
+import org.hibernate.ogm.model.spi.Tuple;
+
+/**
+ * Captures useful data around the state of embeddable objects in tuples
+ * Note that the current implementation is *not* specific to MongoDB on purpose.
+ *
+ * @author Emmanuel Bernard &lt;emmanuel@hibernate.org&gt;
+ */
+public class EmbeddableStateFinder {
+	private final Tuple tuple;
+	private final List<String> columns;
+	private Set<String> nonNullEmbeddables = new HashSet<String>();
+	private Map<String,String> columnToMostOuterNullEmbeddableCache = new HashMap<String, String>();
+
+	public EmbeddableStateFinder(Tuple tuple, TupleContext tupleContext) {
+		this.tuple = tuple;
+		this.columns = tupleContext.getSelectableColumns();
+	}
+
+	/**
+	 * Should only called on a column that is being set to null.
+	 *
+	 * Returns the most outer embeddable containing {@code column} that is entirely null.
+	 * Return null otherwise i.e. not embeddable.
+	 *
+	 * The implementation lazily compute the embeddable state and caches it.
+	 * The idea behind the lazy computation is that only some columns will be set to null
+	 * and only in some situations.
+	 * The idea behind caching is that an embeddable contains several columns, no need to recompute its state.
+	 */
+	public String getMostOuterNullEmbeddable(String column) {
```

Comment:
Ok, took me a bit to wrap my head around it :)

I think it is sound functionally, but that impl is tough to grok. The many inline comments make me think that this single method is too long and too complex.

See https://github.com/gunnarmorling/hibernate-ogm/commit/259e4e9c05bd7128c194ec1cc851a147af0fd2fd how it could be decomposed into several methods which are much friendlier IMO. The method names make most of the inline comments superfluous and let you digest one piece a time.